File: backend/app/core/category.py
"""App category detection and confidence.

Priority is strict and confidence is honest: a label match is materially more
trustworthy than a package-name match, and the calibration card depends on that
distinction being real rather than cosmetic.

Complexity: O(T) token lookups plus O(K) multi-word substring checks, where T is
token count and K a small fixed keyword list. No fuzzy matching here — that
lives in the look-alike domain path where it is actually needed.
"""

from __future__ import annotations

import re

from app.core.rules_loader import Rules, get_rules
from app.models.schemas import CategoryResult

_TOKEN_SPLIT = re.compile(r"[^a-z0-9]+")


def _tokenize(value: str) -> frozenset[str]:
    return frozenset(token for token in _TOKEN_SPLIT.split(value.lower()) if token)


def _match(text: str, rules: Rules) -> str | None:
    if not text:
        return None

    normalized = text.lower()
    tokens = _tokenize(normalized)

    for keyword, category_id in rules.multi_word_keywords:
        if keyword in normalized:
            return category_id

    # Single tokens are checked in sorted order so the result does not depend on
    # dict iteration order when several categories match.
    for token in sorted(tokens):
        category_id = rules.single_word_index.get(token)
        if category_id is not None:
            return category_id

    return None


def detect_category(
    label: str | None,
    package: str | None,
    user_choice: str | None,
    rules: Rules | None = None,
) -> CategoryResult:
    """Resolve the category in strict priority order.

    A user selection always wins. Otherwise the label is tried before the
    package name, because package identifiers are noisier than display names.
    """
    rules = rules or get_rules()

    if user_choice:
        if user_choice not in rules.categories:
            raise ValueError(f"unknown category id: {user_choice}")
        return CategoryResult(id=user_choice, confidence="high", method="user")

    label_match = _match(label or "", rules)
    if label_match is not None:
        return CategoryResult(id=label_match, confidence="medium", method="keyword")

    package_match = _match(package or "", rules)
    if package_match is not None:
        return CategoryResult(id=package_match, confidence="low", method="keyword")

    unknown = rules.categories[rules.unknown_category_id]
    confidence = unknown.confidence or "low"
    return CategoryResult(id=unknown.id, confidence=confidence, method="unknown")


def expected_groups(category_id: str, rules: Rules | None = None) -> frozenset[str]:
    rules = rules or get_rules()
    category = rules.categories.get(category_id)
    return category.expected_groups if category else frozenset()


def mismatch_penalty_disabled(category_id: str, rules: Rules | None = None) -> bool:
    """True for the 'unknown' category, where no mismatch penalty applies."""
    rules = rules or get_rules()
    category = rules.categories.get(category_id)
    if category is None:
        return True
    return category.mismatch_penalty_disabled or category.id == rules.unknown_category_id

File: shared/src/commonMain/kotlin/com/appautopsy/analysis/core/Category.kt
package com.appautopsy.analysis.core

import com.appautopsy.analysis.model.CategoryResult
import com.appautopsy.analysis.model.Confidence
import com.appautopsy.analysis.rules.Rules

/**
 * App category detection and confidence.
 *
 * Priority is strict and confidence is honest: a label match is materially more
 * trustworthy than a package-name match, and the calibration card depends on
 * that distinction being real rather than cosmetic.
 *
 * Complexity: O(T) token lookups plus O(K) multi-word substring checks, where T
 * is token count and K a small fixed keyword list. No fuzzy matching here; that
 * lives in the look-alike domain path where it is actually needed.
 */
object Category {

    private val TOKEN_SPLIT = Regex("[^a-z0-9]+")

    private fun tokenize(value: String): Set<String> =
        TOKEN_SPLIT.split(value.lowercase()).filter { it.isNotEmpty() }.toSet()

    private fun match(text: String, rules: Rules): String? {
        if (text.isEmpty()) return null

        val normalized = text.lowercase()

        // Longest-first, so a specific phrase beats a shorter one inside it.
        for ((keyword, categoryId) in rules.multiWordKeywords) {
            if (normalized.contains(keyword)) return categoryId
        }

        // Sorted so the result does not depend on set iteration order when
        // several categories match the same label.
        for (token in tokenize(normalized).sorted()) {
            rules.singleWordIndex[token]?.let { return it }
        }

        return null
    }

    /**
     * Resolve the category in strict priority order.
     *
     * A user selection always wins. Otherwise the label is tried before the
     * package name, because package identifiers are noisier than display names.
     */
    fun detect(
        label: String?,
        packageName: String?,
        userChoice: String?,
        rules: Rules,
    ): CategoryResult {
        if (!userChoice.isNullOrBlank()) {
            if (userChoice !in rules.categories) {
                throw IllegalArgumentException("unknown category id: $userChoice")
            }
            return CategoryResult(id = userChoice, confidence = Confidence.HIGH, method = "user")
        }

        match(label.orEmpty(), rules)?.let {
            return CategoryResult(id = it, confidence = Confidence.MEDIUM, method = "keyword")
        }

        match(packageName.orEmpty(), rules)?.let {
            return CategoryResult(id = it, confidence = Confidence.LOW, method = "keyword")
        }

        val unknown = rules.categories.getValue(rules.unknownCategoryId)
        return CategoryResult(
            id = unknown.id,
            confidence = Confidence.fromId(unknown.confidence),
            method = "unknown",
        )
    }

    fun expectedGroups(categoryId: String, rules: Rules): Set<String> =
        rules.categories[categoryId]?.expectedGroups ?: emptySet()

    /**
     * True for the `unknown` category, where no mismatch penalty applies.
     *
     * This is the honesty rule: if we could not tell what the app is, we must
     * not punish it for not matching what we could not identify.
     */
    fun mismatchPenaltyDisabled(categoryId: String, rules: Rules): Boolean {
        val category = rules.categories[categoryId] ?: return true
        return category.mismatchPenaltyDisabled || category.id == rules.unknownCategoryId
    }
}